File: rust/tests/message_test.rs
// Message schema tests. A representative subset of the shared corpus
// (spec/corpus/messages.json); the full set is checked cross-language by
// interop/check-messages-rust.sh.

use bonemesh::message::{self, validate};
use serde_json::json;

const MID: &str = "0123456789abcdef0123456789abcdef";

#[test]
fn bmx1_valid() {
    assert_eq!(validate("bmx1", &json!({"t":"bmx1","v":3,"mesh":"acme","e":"AAAA","k":"BBBB","n":"CCCC"})), None);
}
#[test]
fn bmx1_wrong_version() {
    assert_eq!(validate("bmx1", &json!({"t":"bmx1","v":2,"mesh":"acme","e":"AAAA","k":"BBBB","n":"CCCC"})), Some("version"));
}
#[test]
fn bmx1_non_base64() {
    assert_eq!(validate("bmx1", &json!({"t":"bmx1","v":3,"mesh":"acme","e":"not base64!","k":"BBBB","n":"CCCC"})), Some("not-base64"));
}
#[test]
fn envelope_valid() {
    assert_eq!(validate("envelope", &json!({"seq":0,"ct":"3q2+7w=="})), None);
}
#[test]
fn envelope_negative_seq() {
    assert_eq!(validate("envelope", &json!({"seq":-1,"ct":"3q2+7w=="})), Some("seq-range"));
}
#[test]
fn data_valid() {
    assert_eq!(validate("data", &json!({"type":"data","mid":MID,"to":"g","from":"a","ttl":16,"payload":{}})), None);
}
#[test]
fn data_ttl_out_of_range() {
    assert_eq!(validate("data", &json!({"type":"data","mid":MID,"to":"g","from":"a","ttl":256,"payload":{}})), Some("ttl-range"));
}
#[test]
fn data_malformed_mid() {
    assert_eq!(validate("data", &json!({"type":"data","mid":"0123","to":"g","from":"a","ttl":16,"payload":{}})), Some("mid-format"));
    assert_eq!(validate("data", &json!({"type":"data","mid":MID.to_uppercase(),"to":"g","from":"a","ttl":16,"payload":{}})), Some("mid-format"));
}
#[test]
fn ack_valid() {
    assert_eq!(validate("ack", &json!({"type":"ack","mid":MID})), None);
}
#[test]
fn ack_wrong_type() {
    assert_eq!(validate("ack", &json!({"type":"data","mid":MID})), Some("type"));
}

// nak: routed like data, plus hop + reason. The reason value is not enum-checked.
#[test]
fn nak_valid() {
    assert_eq!(validate("nak", &json!({"type":"nak","mid":MID,"hop":"beta","reason":"ttl","to":"alpha","from":"beta","ttl":16})), None);
}
#[test]
fn nak_unknown_reason_valid() {
    assert_eq!(validate("nak", &json!({"type":"nak","mid":MID,"hop":"beta","reason":"some-future-reason","to":"alpha","from":"beta","ttl":16})), None);
}
#[test]
fn nak_wrong_type() {
    assert_eq!(validate("nak", &json!({"type":"data","mid":MID,"hop":"beta","reason":"ttl","to":"alpha","from":"beta","ttl":16})), Some("type"));
}
#[test]
fn nak_bad_mid() {
    assert_eq!(validate("nak", &json!({"type":"nak","mid":"0123","hop":"beta","reason":"ttl","to":"alpha","from":"beta","ttl":16})), Some("mid-format"));
}
#[test]
fn nak_missing_hop() {
    assert_eq!(validate("nak", &json!({"type":"nak","mid":MID,"reason":"ttl","to":"alpha","from":"beta","ttl":16})), Some("missing-field"));
}
#[test]
fn nak_missing_reason() {
    assert_eq!(validate("nak", &json!({"type":"nak","mid":MID,"hop":"beta","to":"alpha","from":"beta","ttl":16})), Some("missing-field"));
}
#[test]
fn nak_ttl_range() {
    assert_eq!(validate("nak", &json!({"type":"nak","mid":MID,"hop":"beta","reason":"ttl","to":"alpha","from":"beta","ttl":0})), Some("ttl-range"));
}

// bye: link-local; reason optional and free-form.
#[test]
fn bye_valid_with_reason() {
    assert_eq!(validate("bye", &json!({"type":"bye","reason":"idle"})), None);
}
#[test]
fn bye_valid_no_reason() {
    assert_eq!(validate("bye", &json!({"type":"bye"})), None);
}
#[test]
fn bye_unknown_reason_valid() {
    assert_eq!(validate("bye", &json!({"type":"bye","reason":"some-future-reason"})), None);
}
#[test]
fn bye_wrong_type() {
    assert_eq!(validate("bye", &json!({"type":"data"})), Some("type"));
}

// Builders round-trip through their own validators.
#[test]
fn builders_produce_valid_messages() {
    assert_eq!(validate("nak", &message::nak(MID, "alpha", "beta", "beta", "ttl", 16)), None);
    assert_eq!(validate("bye", &message::bye(Some("idle"))), None);
    assert_eq!(validate("bye", &message::bye(None)), None);
}

File: php/tests/message.test.php
<?php
// Message-schema validation tests. Reason tags mirror the shared corpus
// (spec/corpus/messages.json). Byte-for-byte agreement is checked by
// bin/interop_checks.php.
use Bonemesh\Message;

const MID = '0123456789abcdef0123456789abcdef';

test('schema verdicts', function () {
    $cases = [
        ['bmx1', ['t' => 'bmx1', 'v' => 3, 'mesh' => 'm', 'e' => 'AA==', 'k' => 'AA==', 'n' => 'AA=='], null],
        ['bmx1', ['t' => 'nope', 'v' => 3, 'mesh' => 'm', 'e' => 'AA==', 'k' => 'AA==', 'n' => 'AA=='], 'type'],
        ['bmx1', ['t' => 'bmx1', 'v' => 2, 'mesh' => 'm', 'e' => 'AA==', 'k' => 'AA==', 'n' => 'AA=='], 'version'],
        ['bmx1', ['t' => 'bmx1', 'v' => 3, 'mesh' => '', 'e' => 'AA==', 'k' => 'AA==', 'n' => 'AA=='], 'empty-mesh'],
        ['bmx1', ['t' => 'bmx1', 'v' => 3, 'mesh' => 'm', 'k' => 'AA==', 'n' => 'AA=='], 'missing-field'],
        ['bmx1', ['t' => 'bmx1', 'v' => 3, 'mesh' => 'm', 'e' => '!!!', 'k' => 'AA==', 'n' => 'AA=='], 'not-base64'],

        ['envelope', ['seq' => 0, 'ct' => 'AA=='], null],
        ['envelope', ['ct' => 'AA=='], 'missing-field'],
        ['envelope', ['seq' => -1, 'ct' => 'AA=='], 'seq-range'],
        ['envelope', ['seq' => 0, 'ct' => '@@@@'], 'not-base64'],

        ['data', ['type' => 'data', 'mid' => MID, 'to' => 'b', 'from' => 'a', 'ttl' => 16, 'payload' => []], null],
        ['data', ['type' => 'x', 'mid' => MID, 'to' => 'b', 'from' => 'a', 'ttl' => 16, 'payload' => []], 'type'],
        ['data', ['type' => 'data', 'mid' => 'short', 'to' => 'b', 'from' => 'a', 'ttl' => 16, 'payload' => []], 'mid-format'],
        ['data', ['type' => 'data', 'mid' => MID, 'from' => 'a', 'ttl' => 16, 'payload' => []], 'missing-field'],
        ['data', ['type' => 'data', 'mid' => MID, 'to' => 'b', 'from' => 'a', 'ttl' => 0, 'payload' => []], 'ttl-range'],
        ['data', ['type' => 'data', 'mid' => MID, 'to' => 'b', 'from' => 'a', 'ttl' => 256, 'payload' => []], 'ttl-range'],

        ['ack', ['type' => 'ack', 'mid' => MID], null],
        ['ack', ['type' => 'nack', 'mid' => MID], 'type'],
        ['ack', ['type' => 'ack', 'mid' => 'NOTHEX0000000000000000000000000x'], 'mid-format'],

        ['nak', ['type' => 'nak', 'mid' => MID, 'hop' => 'b', 'reason' => 'ttl', 'to' => 'a', 'from' => 'b', 'ttl' => 16], null],
        ['nak', ['type' => 'nak', 'mid' => MID, 'hop' => 'b', 'reason' => 'anything-new', 'to' => 'a', 'from' => 'b', 'ttl' => 16], null],
        ['nak', ['type' => 'data', 'mid' => MID, 'hop' => 'b', 'reason' => 'ttl', 'to' => 'a', 'from' => 'b', 'ttl' => 16], 'type'],
        ['nak', ['type' => 'nak', 'mid' => 'short', 'hop' => 'b', 'reason' => 'ttl', 'to' => 'a', 'from' => 'b', 'ttl' => 16], 'mid-format'],
        ['nak', ['type' => 'nak', 'mid' => MID, 'reason' => 'ttl', 'to' => 'a', 'from' => 'b', 'ttl' => 16], 'missing-field'],
        ['nak', ['type' => 'nak', 'mid' => MID, 'hop' => 'b', 'to' => 'a', 'from' => 'b', 'ttl' => 16], 'missing-field'],
        ['nak', ['type' => 'nak', 'mid' => MID, 'hop' => 'b', 'reason' => 'ttl', 'to' => 'a', 'from' => 'b', 'ttl' => 0], 'ttl-range'],

        ['bye', ['type' => 'bye', 'reason' => 'idle'], null],
        ['bye', ['type' => 'bye'], null],
        ['bye', ['type' => 'bye', 'reason' => 'anything-new'], null],
        ['bye', ['type' => 'data'], 'type'],

        ['mystery', [], 'unknown-schema'],
    ];
    foreach ($cases as [$schema, $msg, $want]) {
        assertEq($want, Message::validate($schema, $msg), "$schema " . json_encode($msg));
    }
});

test('builders produce valid messages', function () {
    assertNull(Message::validate('data', Message::data(Message::newMid(), 'a', 'b', Message::DEFAULT_TTL, ['k' => 'v'])));
    assertNull(Message::validate('ack', Message::ack(Message::newMid())));
    assertNull(Message::validate('nak', Message::nak(Message::newMid(), 'a', 'b', 'beta', 'ttl', Message::DEFAULT_TTL)));
    assertNull(Message::validate('bye', Message::bye('idle')));
    assertNull(Message::validate('bye', Message::bye()));
});

test('newMid is well-formed and unique', function () {
    $seen = [];
    for ($i = 0; $i < 1000; $i++) {
        $id = Message::newMid();
        assertEq(1, preg_match('#^[0-9a-f]{32}$#', $id));
        assertTrue(!isset($seen[$id]));
        $seen[$id] = true;
    }
});
